README
================
django-ipgeobase
================

django-ipgeobase - это приложение для определения региона (а также широты и долготы) в России по IP в Django

Установка
=========

Проделываем в командной строке ::

  $ git clone git://github.com/satels/django-ipgeobase.git
  $ cd django-ipgeobase
  $ python setup.py install


Потом следует добавить 'django_ipgeobase' в INSTALLED_APPS и выполнить ::

  $ python manage.py syncdb


Настройка
=========

Необязательные параметры (в settings.py):

* IPGEOBASE_CODING - кодировка этого файла (по-умолчанию windows-1251, можно не менять).
* IPGEOBASE_SOURCE_URL - ссылка на этот файл на сайте-источнике (по-умолчанию установлено, можно не менять)
* IPGEOBASE_SEND_MESSAGE_FOR_ERRORS - отправлять ли сообщения об ошибках на почту при обновлении (по-умолчание, True)


Использование
=============

Для получения объекта ipgeobase (для определения региона) ::

  from django_ipgeobase.models import IPGeoBase

  ip = "212.49.98.48"

  ipgeobases = IPGeoBase.objects.by_ip(ip)
  if ipgeobases.exists():
      ipgeobase = ipgeobases[0]
      print ipgeobase.country #Страна
      print ipgeobase.district #Округ (для указанного ip - Уральский)
      print ipgeobase.region #Регион (Свердловская область)
      print ipgeobase.city #Населенный пункт (Екатеринбург)
      print ipgeobase.ip_block #IP-блок, в который попали (212.49.98.0 - 212.49.98.255)
      print ipgeobase.start_ip, ipgeobase.end_ip #IP-блок в числовом формате
      print ipgeobase.latitude, ipgeobase.longitude #широта и долгота


Обновления базы
---------------

Чтобы обновить базу ipgeobase:

  $ python manage.py ipgeobase_update


CHANGES.txt
История изменений
=================

1.0.6
-----
* Поддержка python3 (python2 тоже поддерживается)
* Поддержка Django > 1.7 (Старые версии тоже поддерживаются)
* Truncate old rows (против Delete раньше)
* Добавлена колонка city_id: Внутренний Ipgeobase id города

1.0.5
-----
* Поддержка python3 (python2 тоже поддерживается)
* Поддержка Django > 1.7 (Старые версии тоже поддерживаются)
* Truncate old rows (против Delete раньше)

1.0.0-beta
-----
Переведено приложение на обновленную базу http://ipgeobase.ru/files/db/Main/geo_files.tar.gz.
Для переход на новую версию нужно выполнить:
./manage.py dbshell
ALTER TABLE django_ipgeobase_ipgeobase add country varchar(16) NOT NULL DEFAULT 'RU';
Изменить колонки district, region, city, latitude, longitude - убрать NOT NULL
Например, для PostgreSQL
ALTER TABLE django_ipgeobase_ipgeobase ALTER district DROP NOT NULL;
ALTER TABLE django_ipgeobase_ipgeobase ALTER region DROP NOT NULL;
ALTER TABLE django_ipgeobase_ipgeobase ALTER city DROP NOT NULL;
ALTER TABLE django_ipgeobase_ipgeobase ALTER latitude DROP NOT NULL;
ALTER TABLE django_ipgeobase_ipgeobase ALTER longitude DROP NOT NULL;
После этого запустить скрипт обновления данных
./manage.py ipgeobase_update

0.5.0
-----
Добавлена возможность отправлять информацию при ошибках выполнения команды обновления базы
Команда скачивания и обновления теперь находятся в одной команде ipgeobase_update

0.4.0
-----
Изменена модель - добавлена ширина и долгота

0.3.0
-----
Добавлен manager по получению объекта IpGeoBase по IP-адресу

0.2.0
-----
Добавлена команда для скачивания архива текстового файла с официального источника и его распаковки

0.1.0
-----
Добавлена команда для конвектирования текстового файла с IP-блоками в таблицу базы данных


django_ipgeobase/models.py
# coding: utf-8
from __future__ import unicode_literals
from django.db import models
from django_ipgeobase.managers import IPGeoBaseManager


class IPGeoBase(models.Model):
    """Таблица перечень блоков ip-адресов с координатами"""

    ip_block = models.CharField(
        verbose_name='Блок IP-адресов',
        max_length=64,
        help_text=(
            "Данное поле состоит из начального и конечного адресов блока, "
            "отделенных друг от друга пробелом, тире и пробелом"
        ),
    )
    start_ip = models.BigIntegerField(
        verbose_name='Начальный IP-адрес блока, преобразованный в число',
        db_index=True,
        help_text=(
            "IP-адрес иммет вид a.b.c.d, где a-d числа в диапазоне 0-255. "
            "Преобразование в число происходит по формуле 256³*a+256²*b+256*c+d"
        ),
    )
    end_ip = models.BigIntegerField(
        verbose_name='Конечный IP-адрес блока, преобразованный в число',
        db_index=True,
    )
    city = models.CharField(
        verbose_name='Город',
        max_length=255,
        null=True,
        db_index=True,
        help_text=(
            "Данное поле содержит в себе название города, соответствующего "
            "данному блоку"
        ),
    )
    city_id = models.IntegerField(
        verbose_name='Ipgeobase Id Города',
        null=True,
        help_text="Внутренний Ipgeobase Id Города",
    )
    region = models.CharField(
        verbose_name='Регион',
        max_length=255,
        null=True,
        help_text=(
            "Данное поле содержит в себе название региона, соответствующего "
            "данному блоку"
        ),
    )
    district = models.CharField(
        verbose_name='Округ',
        max_length=255,
        null=True,
        help_text=(
            "Данное поле содержит в себе название федерального округа, "
            "соответствующего данному блоку IP-адресов"
        ),
    )
    latitude = models.FloatField(verbose_name='Широта', null=True)
    longitude = models.FloatField(verbose_name='Долгота', null=True)
    country = models.CharField(verbose_name='Страна', max_length=16)

    objects = IPGeoBaseManager()

    class Meta:
        index_together = [
            ('start_ip', 'end_ip'),
        ]
        unique_together = [
            ('start_ip', 'end_ip'),
        ]


django_ipgeobase/management/commands/ipgeobase_update.py
# coding: utf-8
from __future__ import print_function, unicode_literals
try:
    from cStringIO import StringIO as BytesIO
except ImportError:
    from io import BytesIO

from django.core.mail import mail_admins
from django.core.management.base import BaseCommand, CommandError
from django.db import connection, transaction
from django_ipgeobase.conf import IPGEOBASE_SOURCE_URL, IPGEOBASE_CODING, \
    IPGEOBASE_SEND_MESSAGE_FOR_ERRORS
try:
    from urllib import urlopen
except ImportError:
    from urllib.request import Request, urlopen as urlopen_py3
    def urlopen(url):
        return urlopen_py3(Request(url))
from zipfile import ZipFile

TRUNCATE_SQL = "TRUNCATE django_ipgeobase_ipgeobase"

INSERT_SQL = """
INSERT INTO django_ipgeobase_ipgeobase
(start_ip, end_ip, ip_block, country, city, region, district, latitude, longitude, city_id)
VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
"""

ERROR_SUBJECT = "Error of command ipgeobase_update"
send_message = IPGEOBASE_SEND_MESSAGE_FOR_ERRORS


class Command(BaseCommand):

    def handle(self, *args, **options):
        print("Download zip-archive...")
        f = urlopen(IPGEOBASE_SOURCE_URL)
        buf = BytesIO(f.read())
        f.close()
        print("Unpacking...")
        zip_file = ZipFile(buf)
        cities_file_read = _read_file(zip_file, 'cities.txt')
        cidr_optim_file_read = _read_file(zip_file, 'cidr_optim.txt')
        zip_file.close()
        buf.close()
        print("Start updating...")
        list_cities = cities_file_read.decode(IPGEOBASE_CODING).split('\n')
        list_cidr_optim = \
            cidr_optim_file_read.decode(IPGEOBASE_CODING).split('\n')
        lines = \
            _get_cidr_optim_with_cities_lines(list_cidr_optim, list_cities)
        cursor = connection.cursor()
        if hasattr(transaction, 'atomic'):
            # django >= 1.6
            try:
                with transaction.atomic():
                    _execute_sql(cursor, lines)
            except Exception as e:
                _handle_error(e)
        else:
            # django < 1.6
            transaction.enter_transaction_management()
            try:
                transaction.managed(True)
                _execute_sql(cursor, lines)
                transaction.commit()
            except Exception as e:
                _handle_error(e)
            finally:
                transaction.rollback()
                transaction.leave_transaction_management()
        return "Table ipgeobase is update.\n"

def _read_file(zip_file, filename):
    try:
        file_read = zip_file.read(filename)
    except KeyError:
        message = "File {} in archive does not found".format(filename)
        if send_message:
            mail_admins(subject=ERROR_SUBJECT, message=message)
        raise CommandError(message)
    return file_read

def _get_cidr_optim_with_cities_lines(list_cidr_optim, list_cities):
    injector = {}
    for line in list_cities:
        row = line.split('\t')
        city_id = row[0]
        injector[city_id] = row[1:]
    for i, line in enumerate(list_cidr_optim):
        row = line.split('\t')
        city_id = row[-1]
        if city_id == '-':
            city_id = None
        row = row[:-1]
        if not row:
            continue
        city_row = injector.get(city_id)
        city_row = city_row or [None]*5
        list_cidr_optim[i] = row + city_row + [city_id]
    return list_cidr_optim

def _execute_sql(cursor, lines):
    print("Delete (truncate) old rows in table ipgeobase...")
    cursor.execute(TRUNCATE_SQL)
    print("Write new data...")
    cursor.executemany(INSERT_SQL, [l for l in lines if l])

def _handle_error(e):
    message = "The data not updated: {}".format(e)
    if send_message:
        mail_admins(subject=ERROR_SUBJECT, message=message)
    raise CommandError(message)


django_ipgeobase/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from django.db import models

try:
    from django.db import migrations  # Django >= 1.7

    class Migration(migrations.Migration):

        initial = True

        dependencies = [
        ]

        operations = [
            migrations.CreateModel(
                name='IPGeoBase',
                fields=[
                    ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                    ('ip_block', models.CharField(help_text='Данное поле состоит из начального и конечного адресов блока, отделенных друг от друга пробелом, тире и пробелом', max_length=64, verbose_name='Блок IP-адресов')),
                    ('start_ip', models.BigIntegerField(db_index=True, help_text='IP-адрес иммет вид a.b.c.d, где a-d числа в диапазоне 0-255. Преобразование в число происходит по формуле 256³*a+256²*b+256*c+d', verbose_name='Начальный IP-адрес блока, преобразованный в число')),
                    ('end_ip', models.BigIntegerField(db_index=True, verbose_name='Конечный IP-адрес блока, преобразованный в число')),
                    ('city', models.CharField(db_index=True, help_text='Данное поле содержит в себе название города, соответствующего данному блоку', max_length=255, null=True, verbose_name='Город')),
                    ('region', models.CharField(help_text='Данное поле содержит в себе название региона, соответствующего данному блоку', max_length=255, null=True, verbose_name='Регион')),
                    ('district', models.CharField(help_text='Данное поле содержит в себе название федерального округа, соответствующего данному блоку IP-адресов', max_length=255, null=True, verbose_name='Округ')),
                    ('latitude', models.FloatField(null=True, verbose_name='Широта')),
                    ('longitude', models.FloatField(null=True, verbose_name='Долгота')),
                    ('country', models.CharField(max_length=16, verbose_name='Страна')),
                ],
            ),
            migrations.AlterUniqueTogether(
                name='ipgeobase',
                unique_together=set([('start_ip', 'end_ip')]),
            ),
        ]

except ImportError:
    # Django < 1.7
    from south.db import db
    from south.v2 import SchemaMigration


    class Migration(SchemaMigration):

        def forwards(self, orm):
            # Adding model 'IPGeoBase'
            db.create_table(u'django_ipgeobase_ipgeobase', (
                (u'id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
                ('ip_block', self.gf('django.db.models.fields.CharField')(max_length=64)),
                ('start_ip', self.gf('django.db.models.fields.BigIntegerField')(db_index=True)),
                ('end_ip', self.gf('django.db.models.fields.BigIntegerField')(db_index=True)),
                ('city', self.gf('django.db.models.fields.CharField')(max_length=255, null=True, db_index=True)),
                ('region', self.gf('django.db.models.fields.CharField')(max_length=255, null=True)),
                ('district', self.gf('django.db.models.fields.CharField')(max_length=255, null=True)),
                ('latitude', self.gf('django.db.models.fields.FloatField')(null=True)),
                ('longitude', self.gf('django.db.models.fields.FloatField')(null=True)),
                ('country', self.gf('django.db.models.fields.CharField')(max_length=16)),
            ))
            db.send_create_signal(u'django_ipgeobase', ['IPGeoBase'])

            # Adding unique constraint on 'IPGeoBase', fields ['start_ip', 'end_ip']
            db.create_unique(u'django_ipgeobase_ipgeobase', ['start_ip', 'end_ip'])


        def backwards(self, orm):
            # Removing unique constraint on 'IPGeoBase', fields ['start_ip', 'end_ip']
            db.delete_unique(u'django_ipgeobase_ipgeobase', ['start_ip', 'end_ip'])

            # Deleting model 'IPGeoBase'
            db.delete_table(u'django_ipgeobase_ipgeobase')


        models = {
            u'django_ipgeobase.ipgeobase': {
                'Meta': {'unique_together': "[('start_ip', 'end_ip')]", 'object_name': 'IPGeoBase'},
                'city': ('django.db.models.fields.CharField', [], {'max_length': '255', 'null': 'True', 'db_index': 'True'}),
                'country': ('django.db.models.fields.CharField', [], {'max_length': '16'}),
                'district': ('django.db.models.fields.CharField', [], {'max_length': '255', 'null': 'True'}),
                'end_ip': ('django.db.models.fields.BigIntegerField', [], {'db_index': 'True'}),
                u'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
                'ip_block': ('django.db.models.fields.CharField', [], {'max_length': '64'}),
                'latitude': ('django.db.models.fields.FloatField', [], {'null': 'True'}),
                'longitude': ('django.db.models.fields.FloatField', [], {'null': 'True'}),
                'region': ('django.db.models.fields.CharField', [], {'max_length': '255', 'null': 'True'}),
                'start_ip': ('django.db.models.fields.BigIntegerField', [], {'db_index': 'True'})
            }
        }

        complete_apps = ['django_ipgeobase']
